src/aggregator/client/queue.go

```diff
@@ -131,92 +139,108 @@ func newInstanceQueue(instance placement.Instance, opts Options) instanceQueue {
+		conn      = newConnection(instance.Endpoint(), connOpts)
+		iOpts     = opts.InstrumentOptions()
+		queueSize = opts.InstanceQueueSize()
 	)
+
+	// Round up queue size to power of 2.
+	// buf is a ring buffer of byte buffers, so it should definitely be many orders of magnitude
+	// below max uint32.
+	qsize := uint32(math.Pow(2, math.Ceil(math.Log2(float64(queueSize)))))
+
 	q := &queue{
-		dropType:           opts.QueueDropType(),
-		log:                iOpts.Logger(),
-		metrics:            newQueueMetrics(iOpts.MetricsScope(), queueSize),
-		instance:           instance,
-		conn:               conn,
-		bufCh:              make(chan protobuf.Buffer, queueSize),
-		doneCh:             make(chan struct{}),
-		maxBatchSize:       maxBatchSize,
-		batchFlushDeadline: writeInterval,
-		buf:                make([]byte, 0, maxBatchSize),
+		dropType: opts.QueueDropType(),
+		log:      iOpts.Logger(),
+		metrics:  newQueueMetrics(iOpts.MetricsScope()),
+		instance: instance,
+		conn:     conn,
+		buf: qbuf{
+			b: make([]protobuf.Buffer, int(qsize)),
+		},
 	}
 	q.writeFn = q.conn.Write
 
-	q.wg.Add(1)
-	go func() {
-		defer q.wg.Done()
-		q.drain()
-	}()
-
 	return q
 }
 
 func (q *queue) Enqueue(buf protobuf.Buffer) error {
-	q.RLock()
-	if q.closed {
-		q.RUnlock()
+	if q.closed.Load() {
 		q.metrics.enqueueClosedErrors.Inc(1)
 		return errInstanceQueueClosed
 	}
-	for {
-		select {
-		case q.bufCh <- buf:
-			q.RUnlock()
-			q.metrics.enqueueSuccesses.Inc(1)
-			return nil
 
-		default:
-			if q.dropType == DropCurrent {
-				q.RUnlock()
-
-				// Close the buffer so it's resources are freed.
-				buf.Close()
-				q.metrics.enqueueCurrentDropped.Inc(1)
-				return errWriterQueueFull
-			}
-		}
+	if len(buf.Bytes()) == 0 {
+		return nil
+	}
+
+	q.mtx.Lock()
+	defer q.mtx.Unlock()
 
-		select {
-		case buf := <-q.bufCh:
-			// Close the buffer so it's resources are freed.
+	if full := q.buf.full(); full {
+		switch q.dropType {
+		case DropCurrent:
+			// Close the current buffer so it's resources are freed.
 			buf.Close()
+			q.metrics.enqueueCurrentDropped.Inc(1)
+			return errWriterQueueFull
+		case DropOldest:
+			// Consume oldest buffer instead.
+			oldest := q.buf.shift()
+			oldest.Close()
 			q.metrics.enqueueOldestDropped.Inc(1)
 		default:
+			return errInvalidDropType
 		}
 	}
+
+	q.buf.push(buf)
+	q.metrics.enqueueSuccesses.Inc(1)
+	return nil
 }
 
 func (q *queue) Close() error {
-	q.Lock()
-	if q.closed {
+	if !q.closed.CAS(false, true) {
 		return errInstanceQueueClosed
 	}
-	q.closed = true
-	q.Unlock()
-
-	close(q.doneCh)
-	q.wg.Wait()
 
-	close(q.bufCh)
 	return nil
 }
 
-func (q *queue) writeAndReset() {
-	if len(q.buf) == 0 {
+func (q *queue) Flush() {
+	q.mtx.Lock()
+
+	if q.buf.size() == 0 {
+		q.mtx.Unlock()
 		return
 	}
-	if err := q.writeFn(q.buf); err != nil {
+
+	buf := _queueConnWriteBufPool.Get().(*[]byte)
+
+	for q.buf.size() > 0 {
+		b := q.buf.shift()
+
+		bytes := b.Bytes()
+		if bytes == nil {
```

Comment:
nit: can you do a `len` check here instead? not sure if it's possible to have a zero-length, non-nil slice returned.

Reply:
yes, good idea, this relies too much on protobuf.Buffers implementation - it doesn't have an easier way to test if it's a zero value.